src/app/InteractionModelEngine.cpp

```diff
@@ -76,21 +78,16 @@ void InteractionModelEngine::Shutdown()
 
     mCommandHandlerList = nullptr;
 
-    //
-    // Since modifying the pool during iteration is generally frowned upon,
-    // I've chosen to just destroy the object but not necessarily de-allocate it.
-    //
-    // This poses a problem when shutting down and restarting the stack, since the
-    // IMEngine is a statically constructed singleton, and this lingering state will
-    // cause issues.
-    //
-    // This doesn't pose a problem right now because there shouldn't be any actual objects
-    // left here due to the synchronous nature of command handling.
-    //
-    // Filed #10332 to track this.
-    //
-    mCommandHandlerObjs.ForEachActiveObject([](CommandHandler * obj) -> bool {
-        obj->~CommandHandler();
+    // Increase magic number to invalid all Handle-s.
+    mMagic++;
+
+    mCommandHandlerObjs.ForEachActiveObject([this](CommandHandler * obj) -> bool {
+        // Modifying the pool during iteration is generally frowned upon, this should not be a issue since we are running in the
+        // main loop.
```

Comment:
The problems with deleting items during iteration wasn't about threading - it was about the state of the data structures being traversed while they were modified during traversal.

Reply:
Comment updated
After reviewing the BitMapStaticPool, it should be fine for modifying the pool during traversal for now, but it is not guranteed.